src/Peptides.cpp

```diff
@@ -16,230 +16,364 @@
 }
 
 
-void Peptides::readFasta(string path) {
-  ifstream fastaIn(path.c_str(),ios::in);
-  string line(""),pep,seq("");
-  unsigned int pepNo=0,proteinNo=0;
-  bool spillOver=false;
-  while (getline(fastaIn,line)) {
-    if (line[0] == '>') {
-      // id line
-      if(spillOver) {
-        assert(seq.size()>0);
-        cleaveProtein(seq,pepNo);
+void Peptides::readFasta(string& path, bool write, std::ostream& os) {
+  ifstream fastaIn(path.c_str(), ios::in);
+  string line, pep, seq;
+  unsigned int pepNo = 0,proteinNo = 0;
+  bool spillOver = false;
+  while (getline(fastaIn, line)) {
+    if (write) os << line << std::endl;
+
+    if (line[0] == '>') { // id line
+      if (spillOver) {
+        assert(!seq.empty());
+        // add peptides and KR-stretches to connectorStrings_
+        cleaveProtein(seq, pepNo);
         seq = "";
       }
       ostringstream newProteinName("");
-      newProteinName << ">" << proteinNamePrefix << ++proteinNo;
-      between.push_back(newProteinName.str());
-      assert(between.size()==pepNo+1);
-    } else {
-      // amino acid sequence
+      newProteinName << ">" << proteinNamePrefix_ << ++proteinNo;
+      connectorStrings_.push_back(newProteinName.str());
+      assert(connectorStrings_.size() == pepNo+1);
+    } else { // amino acid sequence
       seq += line;
       spillOver = true;
     }
-  } 
-  if(spillOver) {
-    cleaveProtein(seq,pepNo);
+  }
+  if (spillOver) {
+    cleaveProtein(seq, pepNo);
+  }
+  if(inferAAFrequency_){
+      background_.setDist(absBackground_.getDist(), replaceI_);
   }
 }
 
 void Peptides::shuffle(const string& in,string& out) {
   out=in;
-  int i = in.length();
-  if (i==0) return;
-  // fisher_yates_shuffle
-  for (; --i; ) {
-    double d=((double)(i+1)*((double)rand()/((double)RAND_MAX+(double)1)));
-    int j=(int)d;
-    if (i!=j){
-      char a = out[i];
-      out[i] = out[j];
-      out[j] = a;
-    }
-  }
+  std::shuffle(out.begin(), out.end(), rGen_);
 }
-void Peptides::mutate(const string& in,string& out) {
+void Peptides::mutate(const string& in, string& out) {
   out=in;
-  int i = in.length();
-  int j=(int)((double)(i+1)*((double)rand()/((double)RAND_MAX+(double)1)));
-  double d=((double)rand()/((double)RAND_MAX+(double)1));
-  out[j] = background.generateAA(d);
+    std::vector<size_t> okIds;
+  if(replaceI_) {
+      okIds = getNotILIdxs(in);
+  } else {
+      okIds = getIdxs(in);
+  }
+  std::size_t i = okIds.size();
+  if (i==0){
+      return;
+  }
+  std::uniform_int_distribution<size_t> distrib(0, i-1);
+  double d= uniformDist();
+  auto mutateId = okIds[distrib(rGen_)];
+  out[mutateId] = background_.generateAA(d);
 }
 
-bool Peptides::recurringPeptide(const string& pep, bool force) {
+bool Peptides::checkAndMarkUsedPeptide(const string& pep, bool force) {
   string checkPep=pep;
-  if (replaceI) {
-    for(string::iterator it=checkPep.begin();it!=checkPep.end();it++) {
-      if (*it=='I')
-        *it='L';
-      if (*it=='N')
```

Comment:
Did you remove the other equivalences because they're not relevant for high-resolution data?